File: lecture_notes/10 refactoring/2-step.py
import json


def pointParse(points):
    return [json.loads(point) if type(point) == str else point for point in points]


def pointsShift(arr, dot_param):
    for dot in arr:
        dot['x'] += dot_param['x']
        dot['y'] += dot_param['y']
    return arr


polyline = [
    {'x': 0, 'y': 0},
    {'x': 10, 'y': 10},
    '{ "x": 20, "y": 20 }',
    {'x': 30, 'y': 30},
]

lineIter = pointParse(polyline)

print(pointsShift(lineIter, {'x': 10, 'y': -5}))

File: lecture_notes/10 refactoring/2_step.js
'use strict';

const shift = (points) => {
  const moved = points.map((point) => (offset) => {
    const type = typeof point;
    if (type === 'object') {
      point.x += offset.x;
      point.y += offset.y;
    } else {
      const i = points.indexOf(point);
      point = JSON.parse(point);
      point.x += offset.x;
      point.y += offset.y;
    }
    return point
  });
  return moved;
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];


const result = shift({ x: 10, y: -5 }, polyline);
console.log({ result });
